### PR TITLE
[dashboard] add license tab to the admin dashboard

### LINKED ISSUE #8413: [licensing] Add a license page to the admin dashboard
Add a page to the admin dashboard that shows the installed license with properties (number allowed users, actual users, domain restriction etc.).

### PR DESCRIPTION
**Pull request by nandajavarma**
## Description

This PR adds a Licence page to the `admin` dashboard. This will help the self-hosted user to understand the features they have available, user limit, etc. The information that are currently available on the UI are:
* Licence level (Community, Professional, or Trial)
* Licence type (Replicated or Gitpod)
* List of features and the ones enabled
* User limit, current user count

Design specs for this can be found under the issue [8413](https://github.com/gitpod-io/gitpod/issues/8413#issuecomment-1089213673)

## Related Issue(s)

Fixes #8413

## How to test

`Admin` > `License` tab has been tested for the following scenarios on a self-hosted setup (it won't be available for SaaS):
- [Replicated community license](https://github.com/gitpod-io/gitpod/tree/main/install/licenses):
  - if they exceeded user limit(lesser features) or is within
- Replicated Professional license:
  - if they exceeded user limit(lesser features) or is within
- Gitpod paid license - [internal](https://www.notion.so/gitpod/Create-a-Gitpod-Self-Hosted-License-fef6b0e3548b4676a9d782f6d5a3df9a):
  - if they exceeded user limit(lesser features) or is within
- Installation with `installer` without a license key (default license)
  - if they exceeded user limit(lesser features) or is within
  
The testing can be done by manually inserting users to the `d_b_user` table of the `mysql` pod to exceed the user limits to test all scenarios.

## Release Notes

```release-note
[dashboard] add license tab to the admin dashboard
```

### CODE REVIEW

**Comment by roboquat** [comment]
@nandajavarma: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/dashboard/src/admin/License.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/License.tsx:57]: **issue**: This is minor and could be ignored for this iteration, but these are fetched from ther backend but feels like there's a typo for some of them. Could we safely rename the following or create follow up issues for cleaning up these?

| BEFORE | AFTER |
|-|-|
| Prebuild | Prebuilds |
| Set Timeout | Timeouts |
| Snapshot | Snapshots |

File: components/dashboard/src/admin/License.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/License.tsx:75]: **question**: Could we align and fix the action button on bottom right area of the card as described in the design specs? Using flex layout could help or simply using some bottom margin here if that sounds safe. Your call. 🏓 

| BEFORE | AFTER |
|-|-|
| button-before | button-after |

File: components/dashboard/src/admin/License.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/License.tsx:54]: **suggestion**: Could we use here normal font size as described in the specs? Going with the `text-base` or removing the `text-sm` above that is inhereted could work. Your call. 🏀

File: components/dashboard/src/admin/License.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/License.tsx:70]: **fyi**: Generally, text-sm should be scarcely used. Defaulting to `text-base` or inheriting the base font size when possible would be ideal. What do you think?

File: components/dashboard/src/admin/License.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/License.tsx:55]: **question**: Probably not the best line to add this comment, but curious what _Paid_ means under the _Professional_ license header. Could we use a more useful subtitle like the following?

> Expires on Dec 22, 2023

Not sure if this information is available though. If not, let's go with the existing chage.

**Review by gtsiolis — COMMENTED** [review]
Thanks for adding this, @nandajavarma! ✨ 

Looks great so far! Added some comments below.

From UX perspective this is getting closer to the finish line. We should probably ask for a review from someone to take a closer look at the code before merging, too. 🏁

File: components/dashboard/src/admin/License.tsx
Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx:75]: @gtsiolis I added users from database without github login, etc. Hence, they are ignored by our users table, it was the best way for me to test.

Comment (gtsiolis) [reply components/dashboard/src/admin/License.tsx:75]: Ah, ok! Sounds good! Resolving this. 📕

File: components/dashboard/src/admin/License.tsx
Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx:55]: This info might be available in certain cases, like a `Trial` Licence from replicated. But yeah might not be available in the case of a professional licence. The amount they paid for the licence(as suggested in your initial design specs) is an information we currently cannot gather from the licence.

Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx:55]: May be in the case of Trial I can try and bring the expiration info, since the account is technically not paid for.

File: components/dashboard/src/admin/License.tsx
Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx:57]: This is a great question! I think I will create another issue for this and deal with it outside the scope, since it might require other code changes in the server/licensor components (I am getting the feature names from an enum in licensor). I don't know if I want to hack more on the front end for this.

File: components/dashboard/src/admin/License.tsx
Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx:171]: Yes exactly! I tried to make it work, but ended up becoming a bit hard to achieve with my very limited front end knowledge. I can move this card component next to the `SelectableCard` components and make a note to standardize in the future!

File: components/dashboard/src/admin/License.tsx
Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx:16]: you can find it here https://gitpod.nanpod.fun/admin/license, I have granted you `admin` role

Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx]: Ah okay, I assumed we use icons. Now that you mention it, I think it might make it much simpler to use HTML codes for check and cross. I have changed the code to reflect this. LMK how the new changes look in design https://gitpod.nanpod.fun/admin/license and https://nvn-license-dash.staging.gitpod-dev.com/admin/license

Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx]: `x` is not being used in the latest update, instead the HTML codes `&#10003;` and `&#10007;` are being used. eg: https://gitpod.nanpod.fun/admin/license, https://nvn-license-dash.staging.gitpod-dev.com/admin/license

File: components/dashboard/src/admin/License.tsx
Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx]: Agreed! updating!

File: components/dashboard/src/admin/License.tsx
Comment (gtsiolis) [reply components/dashboard/src/admin/License.tsx]: > I assumed we use icons.

Most of the times we go with a custom icon. ✔️ 

**issue**: Looks good overall on macOS but I'm concerned that going with this could slightly affect UX as rendering could look quite different on different browsers and operating systems. Do you know how this will render in Linux a system?

**suggestion**: Sounds safer to go with a custom icon taking also into consideration the increased complexity for supporting different icons for light and dark theme. Maybe we could also simply use the `dark:filter-invert` class which we're using elsewhere. Sorry for the back and forth, @nandajavarma! What do you think?

File: components/dashboard/src/admin/License.tsx
Comment (gtsiolis) [reply components/dashboard/src/admin/License.tsx:128]: > Do you suggest we do it?

I'd think yes, following the design specs. In the next iterations we could improve how we surface this information or drop the idea of the colored text following the colored icon.

File: components/dashboard/src/admin/License.tsx
Comment (gtsiolis) [reply components/dashboard/src/admin/License.tsx]: See https://github.com/gitpod-io/gitpod/pull/9343#discussion_r854079292.

File: components/dashboard/src/admin/License.tsx
Comment (gtsiolis) [reply components/dashboard/src/admin/License.tsx:73]: **suggestion**: Since we're here could we also move the checkmark to that the right margin matches the right margin of the button at the bottom of this card? 😇

File: components/dashboard/src/admin/License.tsx
Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx]: updated it to reflect expires value for trial license

File: components/dashboard/src/admin/License.tsx
Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx:74]: I have made the changes!

File: components/dashboard/src/admin/License.tsx
Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx]: done

File: components/dashboard/src/admin/License.tsx
Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx]: moved [`Card` component to a file ](https://github.com/gitpod-io/gitpod/pull/9343/files#diff-6873878bfc81131ca2d4a0cc6e3165f6ad1bbefc418cfa89a0df56ba5e07b4beR1)

File: components/dashboard/src/admin/License.tsx
Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx]: agree! I have added the icons back! changed the `Check` icon to one you suggested. It should be [pretty uniform now.](https://nvn-license-dash.staging.gitpod-dev.com/admin/license)

File: components/dashboard/src/components/SolidCard.tsx
Comment (gtsiolis) [inline components/dashboard/src/components/SolidCard.tsx:11]: **suggestion**: Skipped to mentioned this change in https://github.com/gitpod-io/gitpod/pull/9343#discussion_r853372863. 😅 

```suggestion
                "flex rounded-xl text-base w-72 h-64 px-4 bg-gray-100 dark:bg-gray-800 text-gray-600 dark:text-gray-600" +
```

File: components/dashboard/src/admin/License.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/License.tsx:152]: **nitpick**: Adding below a different green color to use on the dark theme following the initial specs.

```suggestion
                <div className="text-green-600 dark:text-green-400">You have an active professional license.</div>
```

File: components/dashboard/src/admin/License.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/License.tsx:70]: **nitpick**: Moving this back would be great! Sorry if the suggestion in https://github.com/gitpod-io/gitpod/pull/9343#discussion_r853460822 was confusing, @nandajavarma! I meant to keep the top margin and padding as is but move the button after the text elements.

```suggestion
                        <div className="text-gray-600 dark:text-gray-200 py-2 flex-row flex font-semibold items-center">
```

File: components/dashboard/src/admin/License.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/License.tsx:86]: **issue**: Maybe it's worth opening this in a new tab since it's a link outside the dashboard. WDYT?

Eventually we could improve how we differentiate external links (see https://github.com/gitpod-io/gitpod/issues/7232) but that's a different discussion and out of the scope of this PR.

File: components/dashboard/src/admin/License.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/License.tsx:185]: **question**: This is minor but could we safely use the same color for the icon as the text next to the icon here as seen in the specs?

- For the gray version, we could use `gray-600` for the light theme and `gray-50` for the dark theme.
- For the green version, we could use `green-600` for the light theme and `green-400` for the dark theme.
- For the red version, we could use  `red-700` for the light theme and `red-400` for the dark theme.

File: components/dashboard/src/admin/License.tsx
Comment (corneliusludmann) [inline components/dashboard/src/admin/License.tsx:87]: @gtsiolis Don't we also have a link that looks like a button (`a` tag with `href` etc)? Since this is just a link to another page, in my opinion that would be preferable so that we can get rid of the `onClick` JavaScript event.

Advantages:
- Users can see the link target on hover.
- Users can open the page in a new tab.

File: components/dashboard/src/admin/License.tsx
Comment (corneliusludmann) [inline components/dashboard/src/admin/License.tsx:48]: NIT: tiny typo
```suggestion
                subtitle="License associated with your Gitpod installation"
```

File: components/dashboard/src/admin/License.tsx
Comment (corneliusludmann) [inline components/dashboard/src/admin/License.tsx:38]: NIT: Shouldn't it mean “no user limit”? :thinking: 
```suggestion
    // if user seats is 0, it means that there is no user limit in the license
```

**Review by gtsiolis — COMMENTED** [review]
Thanks for the ping, @nandajavarma! 🏓 

This is almost at the finish line! 🏁 

Left a few more comments below.

File: components/dashboard/src/admin/License.tsx
Comment (corneliusludmann) [inline components/dashboard/src/admin/License.tsx:141]: In this code, the `validUntil` param is used for Trial licenses only. However, every license could have a limited validation (I expect that we usually issue 1-year licenses that need to be renewed). Right?

**Review by corneliusludmann — COMMENTED** [review]
Thanks for the great work, Nandaja! :rocket: 

I left a few comments.

Together with the flow chart @lucasvaltl is working on about the different states the license could reach (valid, expired, too many users, wrong domain, etc.) we need to think about how we could test every case (unit tests?) and how the license page actually looks like for each case. But that's follow-up work.

File: components/dashboard/src/admin/License.tsx
Comment (corneliusludmann) [reply components/dashboard/src/admin/License.tsx:87]: Related: https://github.com/gitpod-io/gitpod/pull/9343#discussion_r855126635

**Comment by gtsiolis** [comment]
> ... we need to think about how we could test every case (unit tests?) and how the license page actually looks like for each case. But that's follow-up work.

@corneliusludmann Triggered also from the comment in https://github.com/gitpod-io/gitpod/pull/9343#discussion_r853845336 (Cc @nandajavarma), I'd also like to better understand the options that exist and the matrix of possible use cases depending on which option a user picks to install Gitpod, so that we make the license page flexible to support all cases. Please, keep me in the loop or ping me in a follow up issue. 🏓

File: components/dashboard/src/admin/License.tsx
Comment (gtsiolis) [reply components/dashboard/src/admin/License.tsx:87]: Thanks for noticing, @corneliusludmann! The intent of using a primary (confirm) action button here was to provide at least one primary action users can take. However, since there is no self-serve flow (yet) where users can request or manage their license, going with a link sounds good.

We could keep the button as is and introduce an appending icon to indicate the external link or simply go with a plain `<a>` link. Hint: Using the `gp-link` class could help as we need a different blue color for the light and dark theme.

| Button | Link (Light) | Link (Dark) |
|-|-|-|
| Community | Community | Community-1 |


I've also updated the specs linked in https://github.com/gitpod-io/gitpod/issues/8413#issuecomment-1089213673 to include these new designs. Cc @nandajavarma

**Comment by jldec** [comment]
Hi, would you mind making me (jldec) admin on the preview env so that I can look at the admin UI.
(happy to follow instructions if there is a way for me to do this myself)

**Comment by corneliusludmann** [comment]
> Hi, would you mind making me (jldec) admin on the preview env so that I can look at the admin UI. (happy to follow instructions if there is a way for me to do this myself)

@jldec Open a Gitpod workspace from this PR and run `leeway run components:make-user-admin` ([source](https://www.notion.so/gitpod/How-we-develop-e8ce7e562478458a9223eb9b797415b2#74153318dfc74fb589d8ac12c080a187)).

**Comment by jldec** [comment]
Here's what I see when I look at the preview env (I have narrowed my browser viewport). The comments are just first impressions to someone non-familiar with all the details - not intended as criticism 🙏. Generally speaking I do like the look of the design.

Screenshot 2022-04-21 at 21 59 59


- The 2 boxes are not responsive, and don't wrap when the viewport is narrower - I would have expected wrapping behavior given the design (cc: @gtsiolis)

- The baseline of the X's in the left box are not aligned with the text next to them 

- The vertical position of the "Compare Plans" link seems too high, leaving a large space between the text and the bottom of the box.

- (suggestion) I assume that the X's in the left box mean that the features are unavailable.  This is slightly unclear because some web forms show checked checkboxes with an X and unchecked checkboxes as empty but with the checkbox frame. It would be better if there were at least one item with a checkbox, to make it easier to understand the meaning of the X vs. check iconography. (cc: @gtsiolis)

- I interpret the X's in the following way: _all users have lost access to licensed features because the number of seats was exceeded_ -- Is this the intended interpretation?

- (suggestion) Is the 6/5 exceeded usage limit message correct? I guess this is meant to convey that this instance has a professional licence with 5 seats, and 6 users have registered. It would be clearer if the 5 purchased seats were also explicitly mentioned together with the "Professional" license description on the left, since that feels like one of the most important details. 

- I don't see the "Gitpod" License type in the [linked self-hosted license page](https://www.gitpod.io/self-hosted)) and can't really interpret what that means.

**Comment by jldec** [comment]
I was going through another [admin epic](https://github.com/gitpod-io/gitpod/issues/7597), and noticed that the Admin Dashboard should no longer be limited to license holders. This means that Admin Dashboard could always be shown as checked in the left box :)

See https://github.com/gitpod-io/gitpod/pull/8593/commits/ebb90c600efd7adf72c087260cb0ef40648d2e89 and https://github.com/gitpod-io/gitpod/issues/6932#issuecomment-1062837883 - cc: @MrSimonEmms

**Comment by nandajavarma** [comment]
@jldec The bit about `admin` dashboard always being shown is only true in the case of `community` license. As in, a person setting up self-hosted Gitpod with a `Starter` plan(until 10 users they get all the enterprise features) gets a fallback upon crossing the 10 user limit and they get only `admin` dashboard. This is not the case with a paid license. We do not offer fallback. We disable signups and do not offer any of the features(even admin dashboard).

**Comment by corneliusludmann** [comment]
Whether the admin dashboard is available or not is outside of the scope of this pull request, in my opinion. We need to fix this on the license code level (if needed) and not in the UI.

Nevertheless, good heads up, @jldec! :pray: 

---

@lucasvaltl We need to consider this in our license evaluation work and check implementation vs. what we really want. Homepage tells currently that admin dashboard is a EE feature but I think Jürgen is right here that the admin dashboard is intentional always available (code-wise).

File: components/dashboard/src/admin/License.tsx
Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx]: Sorry about the confusion @gtsiolis , I believe I have fixed this [now](https://nvn-license-dash.staging.gitpod-dev.com/admin/license)

File: components/dashboard/src/admin/License.tsx
Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx]: I have changed it to be a link instead of a button according to your designs.

File: components/dashboard/src/admin/License.tsx
Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx]: done!

File: components/dashboard/src/admin/License.tsx
Comment (gtsiolis) [inline components/dashboard/src/admin/License.tsx:82]: **suggestion**: What do you think of using text-sm here or the parent <a> element so that link text is not larger than the rest of the information on this card as also match the line height to the height of the icon used, `20px`?

```suggestion
                            <span className="text-sm">Compare Plans</span>
```

| BEFORE | AFTER |
|-|-|
| link-before | link-after |

**Review by gtsiolis — APPROVED** [review]
Thanks @nandajavarma for tackling all pending discussions!

FWIW, regarding responsive layout mentioned in https://github.com/gitpod-io/gitpod/pull/9343#issuecomment-1111267703 is part of a larger topic affecting more parts of the dashboard, see https://github.com/gitpod-io/gitpod/issues/4050. In case there's anything we could do ad-hoc to improve the responsive layout in every PR that introduces new components or patterns we could do that! However, it's also ok or even preferred to leave this out of the scope of this PR.

UX in this PR now looks mergable! 🛹 

Do you think it's worth passing this to someone else to take a closer look at the code changes? 🏀

File: components/dashboard/src/admin/License.tsx
Comment (nandajavarma) [reply components/dashboard/src/admin/License.tsx]: love it!

**Review by jldec — APPROVED** [review]
Thank you @nandajavarma - this looks much more polished now.
You have addressed all of my feedback in your comments or in fixes.
LGTM

File: components/dashboard/src/admin/License.tsx
Comment (lucasvaltl) [reply components/dashboard/src/admin/License.tsx]: Hooking into this conversation: It would be great if we can match the naming here to what is on the pricing page - e.g. it’s workspace sharing here in the admin UI but shared workspaces in the pricing page. Not urgent but makes things tidier.

Comment (gtsiolis) [reply components/dashboard/src/admin/License.tsx]: Opened https://github.com/gitpod-io/gitpod/issues/9785 to track this discussion and avoid discussions on closed PRs. 🏓 Cc @nandajavarma @lucasvaltl

**Pull request merged**
